src/details/ArborX_DetailsBufferOptimization.hpp

```diff
@@ -70,103 +69,27 @@ struct InsertGenerator
                                    std::is_same<V, SpatialPredicateTag>{}>
   operator()(int predicate_index, int primitive_index) const
   {
-    auto const permuted_predicate_index = _permute(predicate_index);
-    // With permutation, we access offset in random manner, and
-    // _offset(permutated_predicate_index+1) may be in a completely different
-    // place. Instead, use pointers to get the correct value for the buffer
-    // size. For this reason, also take a reference for offset.
-    auto const &offset = _offset(permuted_predicate_index);
-    auto const buffer_size = *(&offset + 1) - offset;
     auto &count = _counts(predicate_index);
 
     _callback(Access::get(_permuted_predicates, predicate_index),
               primitive_index, [&](ValueType const &value) {
                 int count_old = Kokkos::atomic_fetch_add(&count, 1);
-                if (count_old < buffer_size)
-                  _out(offset + count_old) = value;
+                _out.insert(Kokkos::pair<int, int>{predicate_index, count_old},
```

Comment:
Need to check the return of `insert`, as it may fail. May want to shortcut in that case.